http-clients/src/main/java/com/palantir/remoting3/clients/UserAgent.java

```diff
@@ -0,0 +1,123 @@
+
+    @Value.Derived
+    default String headerFormat() {
+        return Joiner.on(", ").join(Lists.transform(agents(), Agent::headerFormat));
+    }
+
+    static UserAgent of(String serviceName, String instanceId, String version) {
+        Agent agent = ImmutableAgent.builder()
+                .serviceName(serviceName)
+                .instanceId(instanceId)
+                .version(version)
+                .build();
+        return ImmutableUserAgent.builder()
+                .addAgents(agent)
+                .build();
+    }
+
+    static UserAgent of(String serviceName, String version) {
+        Agent agent = ImmutableAgent.builder()
+                .serviceName(serviceName)
+                .version(version)
+                .build();
+        return ImmutableUserAgent.builder()
+                .addAgents(agent)
+                .build();
+    }
+
+    static UserAgent append(UserAgent userAgent, String serviceName, String version) {
+        Agent agent = ImmutableAgent.builder()
+                .serviceName(serviceName)
+                .version(version)
+                .build();
+        return ImmutableUserAgent.builder()
+                .from(userAgent)
+                .addAgents(agent)
+                .build();
+    }
+
+    @Value.Immutable
+    @ImmutablesStyle
+    interface Agent {
+        Pattern SERVICE_NAME_REGEX = Pattern.compile("([a-zA-Z][a-zA-Z0-9\\-]*)");
+        Pattern INSTANCE_ID_REGEX = SERVICE_NAME_REGEX;
+        // See https://github.com/palantir/sls-packaging/blob/develop/sls-versions/src/main/java/com/palantir/slspackaging/versions/SlsProductVersions.java
+        Pattern[] ORDERABLE_VERSION = new Pattern[]{
+                Pattern.compile("^[0-9]+\\.[0-9]+\\.[0-9]+-[0-9]+-g[a-f0-9]+$"),
+                Pattern.compile("^[0-9]+\\.[0-9]+\\.[0-9]+$"),
+                Pattern.compile("^[0-9]+\\.[0-9]+\\.[0-9]+-rc[0-9]+$"),
+                Pattern.compile("^[0-9]+\\.[0-9]+\\.[0-9]+-rc[0-9]+-[0-9]+-g[a-f0-9]+$")
+        };
+
+        String serviceName();
+        Optional<String> instanceId();
+        String version();
+
+        @Value.Lazy
+        default String headerFormat() {
+            StringBuilder builder = new StringBuilder(serviceName());
+
+            instanceId().ifPresent(id -> builder.append("/").append(id));
+
+            builder.append(" (").append(version()).append(")");
+            return builder.toString();
+        }
+
+        @Value.Check
+        default void check() {
+            checkArgument(SERVICE_NAME_REGEX.matcher(serviceName()).matches(),
+                    "Illegal service name format: %s", serviceName());
+            if (instanceId().isPresent()) {
+                checkArgument(INSTANCE_ID_REGEX.matcher(instanceId().get()).matches(),
+                        "Illegal instance id format: %s", instanceId().get());
+            }
+            checkArgument(isOrderableVersion(version()), "Illegal version format: %s", version());
```

Comment:
calling out that depending on where/how this class will get used, failing here might to too harsh and it could be more appropriate to log a warning and fill in a dummy version (0.0.0).

Reply:
I think this will break integration tests that run with versions of 1.0.0-1-g123abc since that's not orderable.  So this wouldn't be usable in integration tests